tests/unit/websocket.cpp

```diff
@@ -0,0 +1,298 @@
+    EXPECT_NO_THROW(client.Write(R"({"username": "user" "password": "123"})"));
+    const auto response = nlohmann::json::parse(client.Read());
+    const auto message_header = response[kResponseMessage].get<std::string>();
+    const auto message_first_part = message_header.substr(0, message_header.find(": "));
+
+    EXPECT_FALSE(response[kResponseSuccess]);
+    EXPECT_EQ(message_first_part, auth_fail) << "Message was: " << message_header;
+  }
+}
+
+TEST_F(WebSocketServerTest, WebsocketAuthenticationWhenAuthPasses) {
+  constexpr auto auth_success = R"({"message":"User has been successfully authenticated!","success":true})";
+
+  {
+    auto client = Client();
+    EXPECT_NO_THROW(client.Connect(ServerAddress(), ServerPort()));
+    EXPECT_NO_THROW(client.Write(R"({"username": "user", "password": "123"})"));
+    const auto response = client.Read();
+
+    EXPECT_TRUE(response == auth_success) << "Response was: " << response;
+  }
+}
+
+TEST_F(WebSocketServerTest, WebsocketAuthenticationWithMultipleAttempts) {
+  constexpr auto auth_success = R"({"message":"User has been successfully authenticated!","success":true})";
+  constexpr auto auth_fail = "Cannot parse JSON for WebSocket authentication";
+
+  {
+    auto client = Client();
+    EXPECT_NO_THROW(client.Connect(ServerAddress(), ServerPort()));
+    EXPECT_NO_THROW(client.Write(R"({"username": "user" "password": "123"})"));
+    const auto response1 = nlohmann::json::parse(client.Read());
+    const auto message_header1 = response1[kResponseMessage].get<std::string>();
+    const auto message_first_part1 = message_header1.substr(0, message_header1.find(": "));
+
+    EXPECT_FALSE(response1[kResponseSuccess]);
+    EXPECT_EQ(message_first_part1, auth_fail) << "Message was: " << message_header1;
+
+    EXPECT_NO_THROW(client.Connect(ServerAddress(), ServerPort()));
+    EXPECT_NO_THROW(client.Write(R"({"username": "user", "password": "123"})"));
+    const auto response2 = client.Read();
+    EXPECT_TRUE(response2 == auth_success) << "Response was: " << response2;
+  }
+  {
+    auto client1 = Client();
+    auto client2 = Client();
+
+    EXPECT_NO_THROW(client1.Connect(ServerAddress(), ServerPort()));
+    EXPECT_NO_THROW(client2.Connect(ServerAddress(), ServerPort()));
+
+    EXPECT_NO_THROW(client1.Write(R"({"username": "user" "password": "123"})"));
+    EXPECT_NO_THROW(client2.Write(R"({"username": "user", "password": "123"})"));
+
+    const auto response1 = nlohmann::json::parse(client1.Read());
+    const auto message_header1 = response1[kResponseMessage].get<std::string>();
+    const auto message_first_part1 = message_header1.substr(0, message_header1.find(": "));
+    EXPECT_FALSE(response1[kResponseSuccess]);
+    EXPECT_EQ(message_first_part1, auth_fail) << "Message was: " << message_header1;
+
+    const auto response2 = client2.Read();
+    EXPECT_TRUE(response2 == auth_success) << "Response was: " << response2;
```

Comment:
```suggestion
    {
      const auto response = nlohmann::json::parse(client1.Read());
      const auto message_header = response[kResponseMessage].get<std::string>();
      const auto message_first_part = message_header.substr(0, message_header.find(": "));
      EXPECT_FALSE(response[kResponseSuccess]);
      EXPECT_EQ(message_first_part, auth_fail) << "Message was: " << message_header;
    }
    {
      const auto response = client2.Read();
      EXPECT_TRUE(response == auth_success) << "Response was: " << response;
    }
```
And you can also add `SCOPED_TRACE` for more information.

Reply:
I did it deliberately this way to have two sessions simultaneously connected

Reply:
but the sessions will still be active as the clients are not part of the scope.
You just process them in a independent scopes.

Reply:
but if they are in a different scope the client will disconnect, since in the clines destructor we are disconnecting from the server

Reply:
I think we didn't understand each other. I didn't mean to put the clients in a different scope, just the temp variables you used (`response`, `message_header`...) so you don't have to enumerate them. But the current implementation also works!

Reply:
Aha sorry I've overlook that when reading your suggestion